### PR TITLE
Return last operation with list API calls

### PR DESCRIPTION
**Pull request by DanFrei**
## Motivation
Allow clients to obtain the last operation by examining the resource list endpoints in order to construct a meaningful list of results.

## Approach
1. A new named query is introduced to allow performant lookups for the last resource operations
2. Base controller implementation was updated to allow the return of last operations in list APIs
3. The Attach last operations and last operation implementation are unified
4. Added support 4 "in statements" with Named queries 

## Pull Request status
* [x] Initial implementation
* [ ] Refactoring
* [] Unit tests
* [x ] Integration tests

### CODE REVIEW

File: storage/interceptable_repository.go [hunk 1]
```diff
@@ -121,6 +121,10 @@ func (ir *queryScopedInterceptableRepository) QueryForList(ctx context.Context,
 	return ir.repositoryInTransaction.QueryForList(ctx, objectType, queryName, queryParams)
 }
 
+func (ir *queryScopedInterceptableRepository) QueryForListWithInStatement(ctx context.Context, objectType types.ObjectType, queryName NamedQuery, queryParams []interface{}) (types.ObjectList, error) {
```
Comment (alexwo) [inline storage/interceptable_repository.go:124]: remove not used anymore

File: storage/named_query.go [hunk 2]
```diff
@@ -39,6 +41,21 @@ var namedQueries = map[NamedQuery]string{
...
+	FROM operations ops 
+    inner join
+		 (
+			 select max(op.paging_sequence) last_operation_sequence
+			 from operations op
+			 group by resource_id
+		 ) lastOperationPerResource
+		 on lastOperationPerResource.last_operation_sequence = ops.paging_sequence
+	WHERE resource_id in (:id_list)`,
+	QueryByTypeAndIDIn: `
```
Comment (eyalnathan) [inline storage/named_query.go:55]: did you mean to push this query?

File: api/base_controller.go [hunk 3]
```diff
@@ -509,22 +516,76 @@ func cleanObject(object types.Object) {
...
 	if err != nil {
-		return util.HandleStorageError(err, types.OperationType.String())
+		return err
 	}
-	if list.Len() == 0 {
-		log.C(ctx).Debugf("No last operation found for entity with id %s of type %s", objectID, object.GetType().String())
-		return nil
+
+	if lastOperation, ok := ops[objectID]; ok {
+		lastOperation.TransitiveResources = nil
```
Comment (eyalnathan) [inline api/base_controller.go:585]: this is not needed in the list case? why is it done here?

File: application.yml [hunk 4]
```diff
@@ -24,8 +24,7 @@ storage:
   skip_ssl_validation: false
   max_idle_connections: 5
 api:
-  token_issuer_url: http://localhost:8080/uaa
-  client_id: cf
+  token_issuer_url: http://localhost:8080/
```
Comment (eyalnathan) [inline application.yml:27]: why is this changed?

File: test/query_test/query_test.go [hunk 5]
```diff
@@ -60,6 +177,16 @@ var _ = Describe("Service Manager Query", func() {
 			id2 = createNotification(repository, now.Add(-30*time.Minute))
 		})
 
+		It("finds objects by ID in", func() {
```
Comment (eyalnathan) [inline test/query_test/query_test.go:180]: did you mean to push this?

File: test/service_instance_and_binding_test/service_binding_test/service_binding_test.go [hunk 6]
```diff
@@ -224,6 +224,28 @@ var _ = DescribeTestsFor(TestCase{
 			})
 
 			Describe("GET", func() {
+
+				Context("when attach_last_operations is truthy", func() {
```
Comment (eyalnathan) [inline test/service_instance_and_binding_test/service_binding_test/service_binding_test.go:228]: shouldn't this be part of the generic tests for all types? it doesn't have any binding specific logic

File: application.yml (hunk 4 above)
Comment (alexwo) [reply application.yml]: Should not change, reverted it.

File: test/service_instance_and_binding_test/service_binding_test/service_binding_test.go (hunk 6 above)
Comment (DanFrei) [reply test/service_instance_and_binding_test/service_binding_test/service_binding_test.go]: Fixed

File: storage/named_query.go [hunk 7]
```diff
@@ -5,6 +5,8 @@ type NamedQuery int
 const (
 	QueryByMissingLabel NamedQuery = iota
 	QueryByExistingLabel
+	QueryForLastOperationsPerResource
+	QueryByTypeAndIDIn
```
Comment (eyalnathan) [inline storage/named_query.go:9]: this should also be removed

File: test/list.go [hunk 8]
```diff
@@ -486,6 +488,75 @@ func DescribeListTestsFor(ctx *common.TestContext, t TestCase, responseMode Resp
...
+				Context("when the last operation does not exist", func() {
+					It("should return the resource without it", func() {
+						resp := ctx.SMWithOAuth.GET(t.API).
+							WithQuery("attach_last_operations", "true").
+							Expect().Status(http.StatusOK).
+							JSON()
+
+						resource := resp.Path("$.items[0]")
+						lastOp := resource.Object().Value("last_operation")
+						resourceId := lastOp.Object().Value("resource_id").String().Raw()
```
Comment (eyalnathan) [inline test/list.go:522]: why do you need the last operation for the resource ID? it is in $.items[0].id

File: test/list.go [hunk 9]
```diff
@@ -486,6 +488,75 @@ func DescribeListTestsFor(ctx *common.TestContext, t TestCase, responseMode Resp
...
+						criteria := query.ByField(query.EqualsOperator, "resource_id", resourceId)
+						err := ctx.SMRepository.Delete(context.Background(), types.OperationType, criteria)
```
Comment (eyalnathan) [inline test/list.go:524]: the setup (like deleting the last ops) is usually done in BeforeEach, and only the test itself in the It, this way it is much easier to understand what is being tested.

File: test/list.go [hunk 10]
```diff
@@ -486,6 +488,75 @@ func DescribeListTestsFor(ctx *common.TestContext, t TestCase, responseMode Resp
...
+							WithQuery("attach_last_operations", "true").
+							Expect().Status(http.StatusOK).
+							JSON()
+						for _, resource := range resp.Path("$.items[*]").Array().Iter() {
+							currentResourceId := resource.Object().Value("id").String().Raw()
+							currentResourceState := resource.Object().Value("ready").Boolean().Raw()
+							lastOp := resource.Object().Value("last_operation")
+
+							Expect(lastOp.Object().Value("resource_id").String().Raw()).To(Equal(currentResourceId))
+							Expect(lastOp.Object().Value("ready").Boolean().Raw()).To(Equal(currentResourceState))
```
Comment (eyalnathan) [inline test/list.go:504]: what is the point of comparing the "ready" field?

Comment (eyalnathan) [reply test/list.go:504]: and how do we know that this is really the last operation on that resource and not just one of them (not the last)?

File: test/query_test/query_test.go [hunk 11]
```diff
@@ -50,6 +51,81 @@ var _ = Describe("Service Manager Query", func() {
...
+
+			list, err := repository.QueryForList(context.Background(), types.OperationType, storage.QueryForLastOperationsPerResource, queryParams)
+			Expect(err).ShouldNot(HaveOccurred())
+			lastOperation := list.ItemAt(0).(*types.Operation)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(list.Len()).To(BeEquivalentTo(1))
+			Expect(lastOperation.State).To(Equal(types.SUCCEEDED))
+		})
+
+		It("should return the correct last operation after a new one is added to a resource", func() {
```
Comment (eyalnathan) [inline test/query_test/query_test.go:78]:
```
When("new operation is created for the resource", func(){
  BeforeEach - do the setup (add operation)
  It("should return the new operation as last operation", func(){
     verify the expected result
  })
})
```

File: test/query_test/query_test.go [hunk 12]
```diff
@@ -50,6 +51,81 @@ var _ = Describe("Service Manager Query", func() {
...
+			queryParams := map[string]interface{}{
+				"id_list": []string{serviceInstance1.ID},
+			}
+
+			list, err := repository.QueryForList(context.Background(), types.OperationType, storage.QueryForLastOperationsPerResource, queryParams)
+			Expect(err).ShouldNot(HaveOccurred())
+			lastOperation := list.ItemAt(0).(*types.Operation)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(list.Len()).To(BeEquivalentTo(1))
+			Expect(lastOperation.State).To(Equal(types.SUCCEEDED))
```
Comment (eyalnathan) [inline test/query_test/query_test.go:75]: is this enough to be sure that this is the last one? don't you want to compare IDs?

File: test/query_test/query_test.go [hunk 13]
```diff
@@ -50,6 +51,81 @@ var _ = Describe("Service Manager Query", func() {
...
+				Base: types.Base{
+					ID:        "my_test_op_latest",
+					CreatedAt: time.Now(),
+					UpdatedAt: time.Now(),
+					Labels:    make(map[string][]string),
+					Ready:     true,
+				},
+				Description:       "my_test_op_latest",
+				Type:              types.CREATE,
+				State:             types.IN_PROGRESS,
```
Comment (eyalnathan) [inline test/query_test/query_test.go:93]: why IN_PROGRESS? i am worried some maintainer job might try to process it, maybe make it FAILED/SUCCEEDED

File: test/service_instance_and_binding_test/service_binding_test/service_binding_test.go [hunk 14]
```diff
@@ -224,6 +224,7 @@ var _ = DescribeTestsFor(TestCase{
 			})
 
 			Describe("GET", func() {
+
```
Comment (eyalnathan) [inline test/service_instance_and_binding_test/service_binding_test/service_binding_test.go:227]: remove this

File: test/list.go [hunk 15]
```diff
@@ -486,6 +488,75 @@ func DescribeListTestsFor(ctx *common.TestContext, t TestCase, responseMode Resp
...
+								resource.Object().ContainsKey("last_operation")
+							}
+						}
+
+						Expect(foundResource).To(Equal(true))
+					})
+				})
+			}
+
+			Context("when attach_last_operations is falsy", func() {
```
Comment (eyalnathan) [inline test/list.go:547]: this should be in the if.SupportsAsync, otherwise it doesn't test much -  there are no operations anyway, right?

File: storage/named_query.go (hunk 7 above)
Comment (DanFrei) [reply storage/named_query.go]: Done

File: api/base_controller.go [hunk 16]
```diff
@@ -509,22 +516,76 @@ func cleanObject(object types.Object) {
...
+			LastOp.TransitiveResources = nil
+			resource.SetLastOperation(LastOp)
+		}
+	}
+
+	return nil
+}
+
+func getLastOperations(ctx context.Context, resourceIDs []string, repository storage.Repository) (map[string]*types.Operation, error) {
+
```
Comment (eyalnathan) [inline api/base_controller.go:547]: remove?

File: storage/named_query.go [hunk 17]
```diff
@@ -39,6 +40,17 @@ var namedQueries = map[NamedQuery]string{
 	(SELECT ID FROM {{.LABELS_TABLE}} 
 				WHERE key=:key
 				AND {{.ENTITY_TABLE}}.{{.PRIMARY_KEY}} = {{.LABELS_TABLE}}.{{.REF_COLUMN}})`,
+	QueryForLastOperationsPerResource: `
```
Comment (eyalnathan) [inline storage/named_query.go:43]: can we make this one look more like the neighbours? CAPITALS, {{.ENTITY_TABLE}} etc.

File: test/common/common.go [hunk 18]
```diff
@@ -220,6 +220,20 @@ func MapContains(actual Object, expected Object) {
 	}
 }
 
+func RemoveResourceByCriterion(testCtx *TestContext, criterion query.Criterion, resourceType types.ObjectType) error {
+	if err := testCtx.SMRepository.InTransaction(context.TODO(), func(ctx context.Context, storage storage.Repository) error {
```
Comment (eyalnathan) [inline test/common/common.go:224]: Do you need to open the tx explicitly? won't it be created automatically in Delete?

Comment (eyalnathan) [reply test/common/common.go:224]: Sorry, looking in the IDE i would like to rephrase - why not use testCtx.SMRepository.Delete?

Comment (alexwo) [reply test/common/common.go:224]: yes this won't work as there will be an operation missing in the context

File: test/list.go [hunk 19]
```diff
@@ -486,6 +489,130 @@ func DescribeListTestsFor(ctx *common.TestContext, t TestCase, responseMode Resp
 				})
 			}
 
+			if t.SupportsAsyncOperations {
+				Context("when attach_last_operations is truthy", func() {
+					var resourceWithOneOperation, resourceWithFewOperations common.Object
+					var lastOperationForResourceId string
+					var resourceWithOneOperationId, resourceWithFewOperationsId string
+
+					createTestOperation := func(resourceID string, opType types.OperationCategory) (types.Object, string) {
```
Comment (eyalnathan) [inline test/list.go:498]: why the complex return value? seems like only the operation ID is used

File: test/list.go [hunk 20]
```diff
@@ -486,6 +489,130 @@ func DescribeListTestsFor(ctx *common.TestContext, t TestCase, responseMode Resp
...
+						resp := ctx.SMWithOAuth.GET(t.API).
+							WithQuery("attach_last_operations", "true").
+							Expect().Status(http.StatusOK).
+							JSON()
+
+						for _, resource := range resp.Path("$.items[*]").Array().Iter() {
+							itemResourceId := resource.Object().Value("id").String().Raw()
+							if itemResourceId == resourceWithOneOperationId || itemResourceId == resourceWithFewOperationsId {
+								lastOp := resource.Object().Value("last_operation")
+								resource.Object().ContainsKey("last_operation")
```
Comment (eyalnathan) [inline test/list.go:550]: switch order of the lines?

File: storage/named_query.go (hunk 17 above)
Comment (alexwo) [reply storage/named_query.go:43]: Do you see a use case for having ENTITY_TABLE variable in this statement?

File: test/list.go [hunk 21]
```diff
@@ -486,6 +489,130 @@ func DescribeListTestsFor(ctx *common.TestContext, t TestCase, responseMode Resp
...
+
+					It("retrieves the resources list when each resource contains the last operation it is associated to", func() {
+						resp := ctx.SMWithOAuth.GET(t.API).
+							WithQuery("attach_last_operations", "true").
+							Expect().Status(http.StatusOK).
+							JSON()
+
+						for _, resource := range resp.Path("$.items[*]").Array().Iter() {
+							itemResourceId := resource.Object().Value("id").String().Raw()
+							if itemResourceId == resourceWithOneOperationId || itemResourceId == resourceWithFewOperationsId {
```
Comment (eyalnathan) [inline test/list.go:548]: are any other resources expected? if it is always just those 2 the if can be removed

File: test/list.go (hunk 19 above)
Comment (alexwo) [reply test/list.go:498]: yes just copied this from a different test, can be simplified

File: test/list.go (hunk 21 above)
Comment (alexwo) [reply test/list.go:548]: yes.. this test is going to be executed by different tests that prepare content that is not cleaned. actually we can't assume what resources will be stored in the database.

File: test/query_test/query_test.go (hunk 12 above)
Comment (eyalnathan) [reply test/query_test/query_test.go:75]: i am still not sure that this is the right operation.. please don't resolve comments without making any change or leaving a reply

File: test/query_test/query_test.go [hunk 22]
```diff
@@ -50,6 +51,85 @@ var _ = Describe("Service Manager Query", func() {
...
+				Expect(err).ShouldNot(HaveOccurred())
+			})
+			It("should return the new operation as last operation", func() {
+				queryParams := map[string]interface{}{
+					"id_list": []string{serviceInstance1.ID},
+				}
+				list, err := repository.QueryForList(context.Background(), types.OperationType, storage.QueryForLastOperationsPerResource, queryParams)
+				Expect(err).ShouldNot(HaveOccurred())
+				lastOperation := list.ItemAt(0).(*types.Operation)
+				Expect(err).ShouldNot(HaveOccurred())
```
Comment (eyalnathan) [inline test/query_test/query_test.go:109]: not needed

File: api/base_controller.go (hunk 16 above)
Comment (alexwo) [reply api/base_controller.go:547]: getLastOperations is used in attachLastOperations &  attachLastOperation, that's the common code that use's the named query for both methods.

File: test/query_test/query_test.go [hunk 23]
```diff
@@ -50,6 +51,85 @@ var _ = Describe("Service Manager Query", func() {
...
+				Expect(list.Len()).To(BeEquivalentTo(1))
+				Expect(lastOperation.State).To(Equal(types.SUCCEEDED))
+				Expect(list.Len()).To(BeEquivalentTo(1))
+				Expect(lastOperation.ID).To(Equal("my_test_op_latest"))
+			})
+
+		})
+
+		It("should return the last operation for every instances in the query", func() {
```
Comment (eyalnathan) [inline test/query_test/query_test.go:118]: I would add another operation for a different resource, and make sure that i got only the operations for the resources that were requested, and not the other one. I think at the moment this is not tested.

File: test/query_test/query_test.go (hunk 22 above)
Comment (alexwo) [reply test/query_test/query_test.go:109]: here we test that in case a resource has 2 operations (old / new), the query picks up the latest expected one.
should not we test it?

File: storage/named_query.go (hunk 17 above)
Comment (eyalnathan) [reply storage/named_query.go:43]: operations? but i get your point..

File: api/base_controller.go (hunk 16 above)
Comment (eyalnathan) [reply api/base_controller.go:547]: sorry, meant the empty line :)

File: test/query_test/query_test.go (hunk 22 above)
Comment (eyalnathan) [reply test/query_test/query_test.go:109]: Expect(err).ShouldNot(HaveOccurred()) is already done 2 lines above, and there's no new assignment for err, so no point in checking again

File: test/query_test/query_test.go (hunk 12 above)
Comment (alexwo) [reply test/query_test/query_test.go:75]: > i am still not sure that this is the right operation.. please don't resolve comments without making any change or leaving a reply

tests was extended with both:
Expect(lastOperation.State).To(Equal(types.SUCCEEDED))
Expect(lastOperation.Type).To(Equal(types.CREATE))

as we create the instance, we know that it should have exactly one created SUCCEEDED operation

Comment (eyalnathan) [reply test/query_test/query_test.go:75]: the other resource also has a single SUCCEEDED & created operation. you query for a specific resource, it makes sense to verify that you got the operation for that resource and not another. just add `Expect(lastOperation.ResourceID).To(Equal(serviceInstance1.ID))`.

Comment (alexwo) [reply test/query_test/query_test.go:75]: Ahh got it .. i thought you wanted to make sure that it returns that correct "last operation" of the instance in the case of multiple ops for the same resource. I will add an expect to confirm the instance id.. got it 👍
